### PR TITLE
Use hamcrest regex matcher rather than our own

### PR DESCRIPTION
**Pull request by thecoop**
Use the matcher provided by hamcrest rather than our own.

The difference is that our matcher uses `.find()` to search for a regex match anywhere in the string, whereas the hamcrest one uses `.matches()` to check the whole string against the regex. This leads to more specific regex checks.

I've left our own one for YAML tests, as that way we don't need to mangle the regex to add `.*` either side, which might be confusing in test failures.

### CODE REVIEW

File: modules/transport-netty4/src/javaRestTest/java/org/elasticsearch/rest/Netty4BadRequestIT.java [hunk 1]
```diff
@@ -63,7 +64,7 @@ public void testBadRequest() throws IOException {
         );
         assertThat(e.getResponse().getStatusLine().getStatusCode(), equalTo(BAD_REQUEST.getStatus()));
         assertThat(e, hasToString(containsString("too_long_http_line_exception")));
-        assertThat(e, hasToString(matches("An HTTP line is larger than \\d+ bytes")));
+        assertThat(e, hasToString(matchesRegex(Pattern.compile(".*An HTTP line is larger than \\d+ bytes.*", Pattern.DOTALL))));
```
Comment (rjernst) [inline modules/transport-netty4/src/javaRestTest/java/org/elasticsearch/rest/Netty4BadRequestIT.java:67]: Is the exception message not a single line?

**Review by rjernst — APPROVED** [review]
Looks fine. Note though that we have _tons_ of uses of startsWith, I don't think we should change those to use the regex matcher, but I notice at least a couple uses here seem they could use that, and it should be faster than doing a regex.

File: modules/transport-netty4/src/javaRestTest/java/org/elasticsearch/rest/Netty4BadRequestIT.java (hunk 1 above)
Comment (thecoop) [reply modules/transport-netty4/src/javaRestTest/java/org/elasticsearch/rest/Netty4BadRequestIT.java:67]: The message has got some formatted JSON in it, with newlines. `matches()` requires a match against the entire string, and by default `.*` doesn't match newlines.

**Comment by thecoop** [comment]
These all have some kind of regex operator - its not just a string check

**Pull request merged**